docs/pages/pmd/projectdocs/decisions/adr-3.md

```diff
@@ -0,0 +1,198 @@
+  These can change at any time without a MAJOR version change.
+
+### Deprecation and removing of old APIs
+
+* APIs can be deprecated at any time (even in PATCH versions). Deprecated APIs are marked with the
+  `@Deprecated` annotation.
+* Deprecations should be listed in the release notes.
+* Deprecated APIs can only be removed with a MAJOR version change.
+
+### Experimental APIs
+
+* New features often introduce new APIs. These new APIs can be marked with the annotation `@Experimental` at
+  the class or method level.
+* APIs marked with the `@Experimental` annotation are subject to change and are considered **not stable**.
+  They can be modified in any way, or even removed, at any time. You should not use or rely
+  on them in any production code. They are purely to allow broad testing and feedback.
+* Experimental APIs can be introduced or removed with at least a MINOR version change.
+  These experimental APIs should be listed in the release notes.
+* Experimental APIs can be promoted to Public APIs with at least a MINOR version change.
+
+### Guidelines for AST classes
+
+AST classes of the individual language modules are used by custom rule implementations and are considered
+Public API in general. Rules only read the AST and do not need to modify it.
+
+In order to minimize the public API surface of AST classes, the following guidelines apply:
+
+* Concrete AST classes should be final, to avoid custom subclasses.
+* Concrete AST classes should only have a package private constructor to avoid manual instantiation.
+  Only the parser of the language (which lives in the same package) should be able to create new instances
+  of AST classes.
+* Concrete AST classes should not have public setters. All setters should be package private, so that
+  only the parser of the language can call the setters during AST construction.
+
+Non-concrete AST classes (like base classes or common interfaces) should follow similar guidelines:
+* Only package private constructor
+* Only package private setters
+
+### Summary of the annotations
+
+* `@InternalApi` (`net.sourceforge.pmd.annotation.InternalApi`)
+
+  This annotation is used for API members that are not publicly supported API but have to live in
+  public packages (outside `internal` packages).
+  Such members may be removed, renamed, moved, or otherwise broken at any time and should not be
+  relied upon outside the main PMD codebase.
+
+* `@Experimental` (`net.sourceforge.pmd.annotation.Experimental`)
+
+  API members marked with the `@Experimental` annotation at the class or method level are subject to change.
+  It is an indication that the feature is in experimental, unstable state.
+  The API members can be modified in any way, or even removed, at any time, without warning.
+  You should not use or rely on them in any production code. They are purely to allow broad testing and feedback.
+
+* `@ReservedSubclassing` (`net.sourceforge.pmd.annotation.ReservedSubclassing`)
```

Comment:
Maybe we should remove this, as we have AFAIK barely used it. I originally meant to use it to close off the AST types so that they not be extended outside of their module. This is basically a poor man's sealed type.

My suggestion would be to not rely on this annotation, instead just document that AST classes and interfaces are not supposed to be extended, like you did above in this document. Eventually we'll be able to use actual sealed types and have stronger guarantees, but the annotation doesn't guarantee more than documentation.

Reply:
I'm fine with removing this. It's currently not used at all. The less rules, the better, I guess :)